tools/run_tests/xds_k8s_test_driver/framework/test_app/client_app.py

```diff
@@ -314,9 +315,38 @@ def find_active_xds_channel(
             )
 
             try:
-                channel = self.check_channel_successful_calls(
+                channel_first_attempt = self.check_channel_successful_calls(
                     channel, **rpc_params
                 )
+                # Address race where a call to the xDS control plane server has
+                # just started and a channelz request comes in before the call
+                # has had a chance to fail.
+                # With channels to the xDS control plane, the channel can be
+                # READY but the calls could be failing due to failure to fetch
+                # OAUTH2 token. To increase the confidence that we have a valid
+                # channel with working OAUTH2 tokens, we check whether the
+                # channel is in a READY state with active calls twice with an
+                # interval of 2 seconds between the two attempts. If the OAUTH2
+                # token is not valid, the call would fail and be caught in
+                # either the first attempt, or the second attempt. It is
+                # possible that between the two attempts, a call fails and a new
+                # call is started, so we also test for equality between the
+                # started calls of the two channelz results.
+                # There still exists a possibility that a call fails on fetching
+                # OAUTH2 token after 2 seconds (maybe because there is a
+                # slowdown in the system.) If such a case is observed, consider
+                # increasing the interval from 2 seconds to 5 seconds.
+                time.sleep(2)
+                channel_second_attempt = self.check_channel_successful_calls(
```

Comment:
The issue here is that `check_channel_successful_calls` does not refresh the channel, it simply looks at the existing state.

Reply:
I've pushed the fix for it in f163b38, but it turned out to be a bit more involved. @XuanWang-Amos - please review, since I've made changes to this PR too.